spec/config_manager/http_project_config_manager_spec.rb

```diff
@@ -0,0 +1,304 @@
+
+    it 'should not send config update notification when datafile is provided' do
+      notification_center = Optimizely::NotificationCenter.new(spy_logger, error_handler)
+
+      expect(notification_center).not_to receive(:send_notifications)
+
+      Optimizely::HTTPProjectConfigManager.new(
+        sdk_key: 'QBw9gFM8oTn7ogY9ANCC1z',
+        datafile: config_body_JSON,
+        start_by_default: false,
+        notification_center: notification_center
+      )
+    end
+  end
+
+  describe '.Initialize(sdk_key, datafile, auto_update: false)' do
+    it 'should get project config instance' do
+      datafile_project_config = Optimizely::DatafileProjectConfig.new(config_body_JSON, spy_logger, error_handler)
+
+      http_project_config_manager = Optimizely::HTTPProjectConfigManager.new(
+        sdk_key: 'QBw9gFM8oTn7ogY9ANCC1z',
+        datafile: config_body_JSON,
+        auto_update: false,
+        start_by_default: false,
+        logger: spy_logger,
+        error_handler: error_handler
+      )
+
+      http_project_config_manager.start!
+
+      # All instance variables values of http_project_config_manager
+      http_project_config_manager_arr = http_project_config_manager.get_config.instance_variables.map do |attr|
+        http_project_config_manager.get_config.instance_variable_get attr
+      end
+
+      # All instance variables values of datafile_project_config
+      datafile_project_config_arr = datafile_project_config.instance_variables.map do |attr|
+        datafile_project_config.instance_variable_get attr
+      end
+
+      expect(http_project_config_manager_arr).to eql(datafile_project_config_arr)
+    end
+
+    it 'should get instance ready immediately' do
+      start = Time.now
+      http_project_config_manager = Optimizely::HTTPProjectConfigManager.new(
+        sdk_key: 'QBw9gFM8oTn7ogY9ANCC1z',
+        datafile: config_body_JSON,
+        auto_update: false,
+        start_by_default: false
+      )
+      expect(http_project_config_manager.get_config).not_to eq(nil)
+      finish = Time.now
+      expect(finish - start).to be < 1
+    end
+
+    it 'should update config, send config update notification and does not schedule a future update' do
+      notification_center = Optimizely::NotificationCenter.new(spy_logger, error_handler)
+
+      datafile_project_config = Optimizely::DatafileProjectConfig.new(config_body_JSON, spy_logger, error_handler)
+
+      expect(notification_center).to receive(:send_notifications).with(
+        Optimizely::NotificationCenter::NOTIFICATION_TYPES[:OPTIMIZELY_CONFIG_UPDATE]
+      ).once
+
+      http_project_config_manager = Optimizely::HTTPProjectConfigManager.new(
+        sdk_key: 'QBw9gFM8oTn7ogY9ANCC1z',
+        datafile: config_body_JSON,
+        polling_interval: 1,
+        auto_update: false,
+        notification_center: notification_center,
+        logger: spy_logger
+      )
+
+      # Sleep to ensure that new request is not scheduled.
+      sleep 3
```

Comment:
Will this be prone to flakiness and/or incorrect behavior? What if it takes more than 3 secs to schedule the next request?

Reply:
Polling interval has been set to 1. So we can safely assume that it shouldn't take more than 3 seconds to schedule another request. We can increase it to 5 but that increases delay in overall unit test running time.

Reply:
Ok, let's leave it at 3 for now.